### PR TITLE
Skip to latest frame of camera stream by default

### PR DESCRIPTION
**Pull request by sberan**
This makes the `enforce_fps='skip'` behavior the default.

Additionally, we now will by default grab file streams at their native FPS (`file_mode`), skipping frames until we are ready to infer in the same manner as we would when streaming. `enforce_fps` will additionally ensure that each frame of the file is inferred upon by the model.

We detect whether we are working with a file or a stream based on how many frames are available via `CAP_PROP_FRAME_COUNT` at start of the stream. In my testing, this was reliable for streams vs webcams vs static files on mac as well as on the nano with FFMPEG backend. If we are not working with a file, we do not allow `enforce_fps` as we know it will cause problems when the `VideoCapture` process gets too far behind in the stream.

-   [x] New feature (non-breaking change which adds functionality)

## How has this change been tested, please provide a testcase or example of how you tested the change?

Locally (webcam, file, and stream). Also tested streaming on my jetson nano.

### CODE REVIEW

**Comment by yeldarby** [comment]
What happens in cases where inference runs faster than the stream framerate?

**Comment by sberan** [comment]
Great question. There are two cases - streaming vs file mode.

**Streaming:** The stream will block on `.grab()` until there is a frame available. Once a frame is available, it will be retrieved and inference will be triggered.
**File:** After a frame is read, the camera thread will sleep until the next frame should be processed. Once the thread wakes up from sleep, it will read a new frame and inference will be triggered.

In either case, we should be doing the correct thing.

**Comment by yeldarby** [comment]
Does that mean we can only process a video at 1x speed even if our GPU is capable of inferring faster? (That’s probably Ok; other inference mechanisms are probably better for that anyway given all the infra we have built up specifically for async video processing in batch.)

File: inference/core/interfaces/stream/stream.py [hunk 1]
```diff
@@ -70,6 +70,7 @@ def __init__(
         on_prediction: Callable = None,
         on_start: Callable = None,
         on_stop: Callable = None,
+        simulate_process_delay: float = 0,
```
Comment (paulguerrie) [inline inference/core/interfaces/stream/stream.py:73]: Is this just for testing? If so, should we remove it?

**Comment by sberan** [comment]
Yep, I found it handy for testing and thought others might too but can
remove it no problem.

**Comment by yeldarby** [comment]
> I will change this behavior so that it runs video faster than native FPS if inference is faster.

I think this may actually be desirable behavior here. The original reason we added static video processing to this was because a customer wanted to simulate a production environment in a repeatable way.

**Comment by sberan** [comment]
Gotcha, I will leave as is. Maybe `enforce_fps` should run faster than native FPS if possible?

**Comment by yeldarby** [comment]
Yeh that seems to make sense since `enforce_fps` is supposed to run on every frame

**Comment by sberan** [comment]
@yeldarby when enforce_fps is true, we are now pulling frames as fast as we can infer - regardless of native framerate. I also added code to wait until we know what our processing rate will be - previously we were grabbing a bunch of frames at 30 FPS before we've processed our first frame.

@paulguerrie the debug code is removed.

🚀

File: inference/core/interfaces/stream/stream.py (hunk 1 above)
Comment (paulguerrie) [reply inference/core/interfaces/stream/stream.py:73]: @sberan maybe you have an un-pushed commit? Or maybe just missed this piece of the debugging code when removing? :)

**Review by paulguerrie — APPROVED** [review]

**Pull request merged**